file.js

```diff
@@ -65,8 +79,52 @@ module.exports = function (file, block_size, flags) {
         })
       })
     },
+    /**
+     * Writes a buffer directly to a position in the file. This opens the file
+     * with another file descriptor so that the main file descriptor can just
+     * append and read without doing any positional writes.
+     *
+     * @param {buffer} buf - the data to write to the file
+     * @param {number} pos - position in the file to write the buffer
+     * @param {function} cb - callback that returns any error as an argument
+     */
+    write: (buf, pos, cb) => {
+      offset.once((_offset) => {
+        const endPos = pos + buf.length
+        const isPastOffset = endPos > _offset
+
+        if (isPastOffset) {
+          return cb(new Error(`cannot write past offset: ${endPos} > ${_offset}`))
+        }
+
+        let shouldRemove = false
+        const rm = writing(function (_writing) {
+          if (shouldRemove === true) {
+            return rm()
+          }
+
+          if (_writing === true) {
+            return
+          } else {
+            writing.set(true)
+            shouldRemove = true
```

Comment:
these lines should most probably be the other way around.
firstly, calling `writing.set(true)` can trigger listeners, which could call other functions, so basically anything could happen. Then you are setting a state that changes how the thing behaves. In some cases, calling `writing.set()` will interact with `shouldRemove` and having being bitten by this, it's very hard to debug things like this!

Here is a case where I got lucky and figured it out quickly: https://github.com/dominictarr/stream-spec/issues/4

Also the [infamous DAO hack](https://coincentral.com/blockchain-security-vulnerabilities/) also depended on a pattern like this.

Reply:
Hmm, I had it that way originally but it was complaining:

```
/home/christianbundy/src/aligned-block-file/file.js:103
            return rm()
            ^

ReferenceError: rm is not defined
    at Array.<anonymous> (/home/christianbundy/src/aligned-block-file/file.js:103:13)
    at trigger (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:8:34)
    at Function.many.set (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:33:47)
    at /home/christianbundy/src/aligned-block-file/file.js:110:21
    at many (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:23:49)
    at offset.once (/home/christianbundy/src/aligned-block-file/file.js:101:20)
    at Function.many.once (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:39:7)
    at Object.write (/home/christianbundy/src/aligned-block-file/file.js:92:14)
    at Object.write (/home/christianbundy/src/aligned-block-file/blocks.js:188:12)
    at /home/christianbundy/src/aligned-block-file/test/simple.js:102:12
```

Even stranger, checking `typeof rm` was throwing the same issue. The only way I can get it working is to replace `rm()` with `setImmediate(() => rm())`.

Reply:
my hunch is it's the scoping rules of `const` with `var rm` it would work I bet.